### PR TITLE
Modify structure to improve maintainability and add fusion control logic

### PR DESCRIPTION
**Pull request by priseborough**
This started out as work to add GPS innovation checks and fusion control model logic, but during the work it was discovered that the split of functionality between the Ekf and estimator_base classes was making integration of new functions more difficult than desired.

GPS and Baro alt innovation consistency checks have been added.

Fusion control logic has been added to enable automatic transition between GPS and non-GPS operation.

The EstimatorBase class has been made thinner, only contains those functions and variables required to interface to data and parameters and has been renamed the EstimatorInterface class. All other functions and variables have been moved to the Ekf class.

Finally ./Tools/fix_code_style.sh was run on the code to fix style differences.

This branch (and it's  associated Firmware branch) have been test flown. Flight characteristics and innovations were normal and ekf2 processor utilisation was less than 13%.
baro innov
gps pos innov
gps vel innov
gps vert vel innov
magx innov
magy innov
magz innov

### CODE REVIEW

**Comment by tumbili** [comment]
@priseborough @bugobliterator Nice! I will review this.

File: EKF/control.cpp [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	// determine if the vehicle is armed
+	_control_status.flags.armed = _vehicle_armed;
+
+	// record vertical position whilst disarmed to use as a height change reference
+	if (!_control_status.flags.armed) {
+		_last_disarmed_posD = _state.pos(2);
+	}
+
+	// Transition to in-air occurs when armed and when altitude has increased sufficiently from the altitude at arming
+	if (!_control_status.flags.in_air && _control_status.flags.armed && (_state.pos(2) - _last_disarmed_posD) < -1.0f) {
```
Comment (tumbili) [inline EKF/control.cpp]: @priseborough Optionally for this we have a landing detector in PX4. From what I've seen until now it's pretty reliable. The landed flag could be obtained via the base class.

File: EKF/control.cpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+		_control_status.flags.in_air = true;
+	}
+
+	// Transition to on-ground occurs when disarmed.
+	if (_control_status.flags.in_air && !_control_status.flags.armed) {
```
Comment (tumbili) [inline EKF/control.cpp]: @priseborough Also here the landing detector would be a bit more precise.

File: EKF/ekf.cpp [hunk 3]
```diff
@@ -91,29 +137,45 @@ bool Ekf::update()
...
 		_fuse_vert_vel = true;
 		_fuse_hor_vel = true;
 
-	} else if (_time_last_imu - _time_last_gps > 2000000 && _time_last_imu - _time_last_fake_gps > 70000) {
-		// if gps does not update then fake position and horizontal veloctiy measurements
-		_fuse_hor_vel = true;	// we only fake horizontal velocity because we still have baro
-		_gps_sample_delayed.vel.setZero();
+	} else if (!_control_status.flags.gps && !_control_status.flags.opt_flow
+		   && ((_time_last_imu - _time_last_fake_gps > 2e5) || _fuse_height)) {
+		_fuse_pos = true;
```
Comment (tumbili) [inline EKF/ekf.cpp]: @priseborough Would there be any benefit it fusing zero horizontal velocity measurements?

File: EKF/ekf.h [hunk 4]
```diff
@@ -78,11 +80,22 @@ class Ekf : public EstimatorBase
 	// get the diagonal elements of the covariance matrix
 	void get_covariances(float *covariances);
 
+	// ask estimator for sensor data collection decision and do any preprocessing if required, returns true if not defined
+	bool collect_gps(uint64_t time_usec, struct gps_message *gps);
+	bool collect_imu(imuSample &imu);
+
+	filter_control_status_u _control_status = {};
+
+	// get the ekf WGS-84 origin positoin and height and the system time it was last set
```
Comment (tumbili) [inline EKF/ekf.h]: typo

File: EKF/ekf.cpp [hunk 5]
```diff
@@ -91,29 +137,45 @@ bool Ekf::update()
...
+			fuseHeading();
+		}
 	}
 
 	if (_baro_buffer.pop_first_older_than(_imu_sample_delayed.time_us, &_baro_sample_delayed)) {
 		_fuse_height = true;
 	}
 
-	if (_gps_buffer.pop_first_older_than(_imu_sample_delayed.time_us, &_gps_sample_delayed)) {
+	// If we are using GPs aiding and data has fallen behind the fusion time horizon then fuse it
```
Comment (tumbili) [inline EKF/ekf.cpp]: typo GPs

File: EKF/mag_fusion.cpp [hunk 6]
```diff
@@ -44,408 +44,419 @@
...
+			_mag_healthy = false;
+		}
+	}
+
+	if (!_mag_healthy) {
+		return;
+	}
+
+	// update the states and covariance usinng sequential fusion of the magnetometer components
+	for (uint8_t index = 0; index <= 2; index++) {
```
Comment (tumbili) [inline EKF/mag_fusion.cpp]: @priseborough How come you never performed all three fusion steps at once for APM? I thought this has something to do with the computational effort.

**Comment by tumbili** [comment]
@priseborough @bugobliterator Applied to master! Paul, could you please still comment on my comments in this PR?

**Comment by tumbili** [comment]
Not the typos, I fixed those ones.

File: EKF/control.cpp (hunk 1 above)
Comment (priseborough) [inline EKF/control.cpp]: OK - I will have look at it.

File: EKF/control.cpp (hunk 2 above)
Comment (priseborough) [inline EKF/control.cpp]: Noted

File: EKF/ekf.cpp (hunk 3 above)
Comment (priseborough) [inline EKF/ekf.cpp]: My experience with their use on APM was that they did not perform as well for rapid changes in velocity within a small area (indoor flying) which is a typical use case for this mode. Also fusing in last known position made the fallback logic easier to implement. Something we still need to do is define a separate position noise parameter to be used when flying without GPS and tune it to keep the indoor fliers happy.

File: EKF/mag_fusion.cpp (hunk 6 above)
Comment (priseborough) [inline EKF/mag_fusion.cpp]: APM does fuse them on the same prediction step with EKF2. The original EKF spread them across prediction steps due to concerns about timing jitter, however we were able to reduce the fusion time for EKF2. What APM does do is delay the fusion of other measurements until the next frame. This also allows for effective load levelling across multiple instances by offsetting the fusion time horizon by a few frames for each additional instance.

**Pull request merged**